File: mapeable/Mapeable.ComparacionesParaPruebasUnitarias/Manejo de propiedades/BuscadorDePropiedadesPublicas.cs
using System;
using System.Collections.Generic;
using System.Reflection;

namespace ManejoDePropiedades
{
    public class BuscadorDePropiedadesPublicas
    {
        private Type elTipo;
        MemberInfo[] losMiembrosPublicosDeInstancia;

        private IList<Propiedad> lasPropiedades;
        public BuscadorDePropiedadesPublicas(Type elTipo)
        {
            this.elTipo = elTipo;
        }

        public IEnumerable<Propiedad> EncuentreLasPropiedadesPublicas()
        {
            ObtengaLosMiembrosPublicosDeInstancia();
            EncuentreLosQueSonPropiedades();

            return lasPropiedades;
        }

        private void ObtengaLosMiembrosPublicosDeInstancia()
        {
            dynamic losFiltros = BindingFlags.Public | BindingFlags.Instance;
            losMiembrosPublicosDeInstancia = elTipo.GetMembers(losFiltros);
        }

        private void EncuentreLosQueSonPropiedades()
        {
            InicialiceLaListaDePropiedades();
            foreach (MemberInfo miembro in losMiembrosPublicosDeInstancia)
            {
                RegistreElMiembroSiEsUnaPropiedad(miembro);
            }
        }

        private void InicialiceLaListaDePropiedades()
        {
            lasPropiedades = new List<Propiedad>();
        }

        private void RegistreLaPropiedad(PropertyInfo unaPropiedad)
        {
            Propiedad nuevaPropiedad = new Propiedad();
            nuevaPropiedad.Nombre = unaPropiedad.Name;
            nuevaPropiedad.Tipo = unaPropiedad.PropertyType;
            nuevaPropiedad.SePuedeEscribir = unaPropiedad.CanWrite;
            nuevaPropiedad.SePuedeLeer = unaPropiedad.CanRead;
            lasPropiedades.Add(nuevaPropiedad);
        }

        private void RegistreElMiembroSiEsUnaPropiedad(MemberInfo miembro)
        {
            if (ElMiembroEsUnaPropiedad(miembro))
            {
                PropertyInfo laPropiedad = (PropertyInfo)miembro;
                RegistreLaPropiedad(laPropiedad);
            }
        }

        private bool ElMiembroEsUnaPropiedad(MemberInfo miembro)
        {
            return miembro.MemberType == MemberTypes.Property;
        }
    }
}

File: mapeable/Mapeable/Manejo de propiedades/BuscadorDePropiedadesPublicas.vb
Imports System.Reflection

Namespace ManejoDePropiedades
    Public Class BuscadorDePropiedadesPublicas
        Private elTipo As Type
        Dim losMiembrosPublicosDeInstancia() As MemberInfo
        Private lasPropiedades As IList(Of Propiedad)

        Public Sub New(elTipo As Type)
            Me.elTipo = elTipo
        End Sub

        Public Function EncuentreLasPropiedadesPublicas() As IEnumerable(Of Propiedad)
            ObtengaLosMiembrosPublicosDeInstancia()
            EncuentreLosQueSonPropiedades()

            Return lasPropiedades
        End Function

        Private Sub ObtengaLosMiembrosPublicosDeInstancia()
            Dim losFiltros = BindingFlags.Public Or BindingFlags.Instance
            losMiembrosPublicosDeInstancia = elTipo.GetMembers(losFiltros)
        End Sub

        Private Sub EncuentreLosQueSonPropiedades()
            InicialiceLaListaDePropiedades()
            For Each miembro In losMiembrosPublicosDeInstancia
                RegistreElMiembroSiEsUnaPropiedad(miembro)
            Next
        End Sub

        Private Sub InicialiceLaListaDePropiedades()
            lasPropiedades = New List(Of Propiedad)
        End Sub

        Private Sub RegistreLaPropiedad(unaPropiedad As PropertyInfo)
            Dim nuevaPropiedad As New Propiedad
            nuevaPropiedad.Nombre = unaPropiedad.Name
            nuevaPropiedad.Tipo = unaPropiedad.PropertyType
            nuevaPropiedad.SePuedeEscribir = unaPropiedad.CanWrite
            nuevaPropiedad.SePuedeLeer = unaPropiedad.CanRead
            lasPropiedades.Add(nuevaPropiedad)
        End Sub

        Private Sub RegistreElMiembroSiEsUnaPropiedad(miembro As MemberInfo)
            If ElMiembroEsUnaPropiedad(miembro) Then
                RegistreLaPropiedad(miembro)
            End If
        End Sub

        Private Function ElMiembroEsUnaPropiedad(miembro As MemberInfo) As Boolean
            Return miembro.MemberType = MemberTypes.Property
        End Function
    End Class
End Namespace